### ISSUE TITLE
Feature Request: Stream on battery operated cameras

### ISSUE THREAD

**Issue by johannijs** [JurajNyiri/HomeAssistant-Tapo-Control#507]
**Is your feature request related to a problem? Please describe.**
no streaming possible from battery operated camera's

**Describe the solution you'd like**
i was thinking if it's possible to extract the c420 camera steam through google home.
Since I can ask google to play my camera stream on the google-hub, there must be a way to trick
tapo that home assistant is a google cast device?

**Describe alternatives you've considered**
none possible

**Additional context**
only tested on C420, hit maybe works on the doorbell?

**Comment by JurajNyiri** [comment]
This should be possible locally by:
- using the same functionality as we use for downloading recordings but request stream instead
- also since it is a child device, some more changes to request child device stream instead of main hub device
- expose the stream to ha via ffmpeg

This would most likely kill the battery life so we would need to optimise it and do it on demand.

I do not own battery operated camera so we would need help here.

**Comment by JurajNyiri** [comment]
We will need to pass the stream into ffmpeg binary which would pass it to ha similarly like how is it done currently if stream functionality is disabled. As for the battery maybe we just return stream if user requires it? I am not sure if we can detect that we will need to test that.

**Comment by DamienMrtl** [comment]
Now it seems the Google home app on android can also stream the battery cameras, in my case c420. So I tried to scan the network traffic, but it seems the UDP stream is coming from an Amazon compute IP address, so it looks like the stream is not local. Is there any hope to find a solution in this case ?

**Comment by bigverm23** [comment]
I have a C402 and can confirm stream does not work.

**Comment by mateuszdrab** [comment]
I have 2x C400 that also would benefit from an on-demand stream feature.
I noticed the Alexa app is able to stream from them, they connect via the H200 hub.
Happy to help with the development.

**Comment by DamienMrtl** [comment]
I also would like to help in the development of this feature, I have C420 camera. How can we help ?

**Comment by djovic78** [comment]
I've a C410 and would highly appreciate this feature. I could help in debugging.
I can start a stream to my Google Nest Hub devices without any issues. I doubt they get streamed via the cloud? Also the Tapo app obviously can stream as well.

As an alternative: Could a manual recording start and stop be implemented?

**Comment by JurajNyiri** [comment]
Here is code that is able to open stream and save it to file every few seconds for debugging purposes.

https://github.com/JurajNyiri/pytapo/blob/experiment-stream/pytapo/media_stream/streamer.py

Usage:

```
from pytapo import Tapo
from pytapo.media_stream.streamer import Streamer
import asyncio
import os

# mandatory
outputDir = os.environ.get("OUTPUT")  # directory path where videos will be saved
host = os.environ.get("HOST")  # change to camera IP
password_cloud = os.environ.get("PASSWORD_CLOUD")  # set to your cloud password

# optional
window_size = os.environ.get(
    "WINDOW_SIZE"
)  # set to prefferred window size, affects download speed and stability, recommended: 50

print("Connecting to camera...")
tapo = Tapo(host, "admin", password_cloud, password_cloud)


async def download_async():
    print("Getting recordings...")
    downloader = Streamer(tapo, outputDir)
    async for status in downloader.download():
        print(status)
    print("")


loop = asyncio.get_event_loop()
loop.run_until_complete(download_async())

```

Now we need to figure out how to get this in realtime as a stream to HA.

**Comment by JurajNyiri** [comment]
https://github.com/JurajNyiri/pytapo/blob/experiment-stream-hls/experiments/DownloadStream.py

Class can now be started / stopped and works async and creates HLS stream.

**Comment by JurajNyiri** [comment]
Hi @AlexxIT,

I need your help, I've been stuck on this for a while now --- I cannot get audio to work. Tried different codecs, tried remuxing it on a go which I gave up on for now. I think it must be the case of how I am piping it in.

Please see this simple demo:

https://github.com/JurajNyiri/pytapo/blob/experiment-stream-hls/experiments/DownloadStream.py
Using new class https://github.com/JurajNyiri/pytapo/blob/experiment-stream-hls/pytapo/media_stream/streamer.py

Basically it just creates an m3u8 file with current stream using the proprietary Tapo streaming protocol.

I think the reason why it is failing might be because of how I am piping it in.

For downloading of recordings how I solved this was I created a separate file for audio and video, and then merged them:
https://github.com/JurajNyiri/pytapo/blob/experiment-stream-hls/pytapo/media_stream/convert.py#L32 but for obvious reasons that is not going to work for realtime stream. Maybe I can pipe it in from 2 different pipes? But there must be a simpler way. Please let me know your thoughts when you have a bit of time.

**Comment by AlexxIT** [comment]
PCMA not supported in MPEG-TS.That is “proprietary” part of TP-Link. You using HLS as output. Which also using MPEG-TS as file format. You should convert audio to AAC.

**Comment by JurajNyiri** [comment]
@AlexxIT thank you!

I tried that here https://github.com/JurajNyiri/pytapo/commit/cac3207e7b722b336a2f127444d6894b342c6a56 amongs a lot of other changes, like specifying input as alaw, bigger probesize, analyse duration etc etc and still get the same result - `Stream #0:1[0x45], 119, 1/90000: Unknown: none ([144][0][0][0] / 0x0090).`

I wonder, is there no way how to use just a single ffmpeg command to process the stream and do i have to do 2 separate ffmpegs and remux them together like it is done for recordings? Is the command I have right now looking correct to you?

**Comment by AlexxIT** [comment]
I don't know how to pass MPEG-TS with PCMA inside to FFmpeg. I think there is no way. 

You need to write MPEGTS reader. Or search some exist python code. 

I don't know how to pass two pipes to ffmpeg. 
I think you can pass named pipes.

**Comment by JurajNyiri** [comment]
Code in https://github.com/JurajNyiri/pytapo/tree/experiment-stream-hls now creates a separate audio and video stream as well as merges them. Now we need a bit of cleanup and expose it to HA.

**Comment by JurajNyiri** [comment]
Code above was updated to use pipes instead of temporary m3u8 files. This reduces IO significantly.

Delay has been also improved.

How to test:


1. Download https://github.com/JurajNyiri/pytapo/tree/experiment-stream-hls as a zip, extract it
2. Install python3 
3. move https://github.com/JurajNyiri/pytapo/blob/experiment-stream-hls/experiments/DownloadStream.py one directory above
4. Execute it like so

```
OUTPUT=folder_path_where_m3u8_will_be_stored HOST=your_camera_ip PASSWORD_CLOUD=your_tplink_password python3 DownloadStream.py
```
5. Adjust keepRunningFor (inside DownloadStream.py) you moved up to value of 3600 to ensure it keeps running for a bit
6. Open m3u8 generated inside OUTPUT folder with VLC
7. Once verified, kill it by using ctrl+c in terminal

**Comment by djovic78** [comment]
> 5. Adjust keepRunningFor (inside DownloadStream.py) you moved up to value of 3600 to ensure it keeps running for a bit
> 6. Open m3u8 generated inside OUTPUT folder with VLC
> 7. Once verified, kill it by using ctrl+c in terminal

Hi @JurajNyiri. Thanks for your work. I tested it today and it doesn't seem to work for me. I get the following result:

- connection to the camera is established correctly
- ffmpeg starts the stream
- I see the following files in OUTPUT:
- stream_output.m3u8, stream_output0.ts stream_output1.ts stream_output2.ts (eventually stream_output3.ts)
- stream_output.m3u8 grows to approx. 200 B, while the others have about 230KiB
- each file contains about 3 seconds of video
- the stream_output.m3u8 opens in VLC and also shows about 3 seconds of footage then stalls
- last line of stdout, before it stalls:
`{'currentAction': 'Streaming', 'ffmpegLog': '[AVIOContext @ 0x764bdc0010c0] Statistics: 202 bytes written, 0 seeks, 1 writeouts'}`

My Camera is a C410. Streaming in the app or to Google Nest Hub works.

Let me know what I can do to help. My coding skills are a bit rusty though.

**Comment by djovic78** [comment]
Update: Just tried running the DownloadStream.py with ENABLE_AUDIO=yes.

With that enabled the stream is working correctly in the sense that it now constantly updates. 
Some observations:

- It has a delay of 5-10 seconds 
- I do not get any audio 

Both are actually totally fine for my use case.

**Comment by JurajNyiri** [comment]
Thank you for confirming @djovic78 !

Next step would be to update integration to expose the ffmpeg stream somehow. Instead of writing to m3u8, it should return the ffmpeg (it already returns pid) and use it to expose the stream through the camera entity.

Additionally, we will need to change the logic in integration so that this type of stream can be easily enabled / disabled through configuration options and works for cameras without rtsp (currently camera entities are disabled for those).

**Comment by sarkinios** [comment]
@JurajNyiri Thanks a lot for your efforts! I tried running DownloadStream.py, but it gets stuck at {'currentAction': 'Streaming', 'ffmpegLog': "[pipe @ 0x556ca551a800] Setting default whitelist 'crypto,data'"} — no .m3u8 or other output files are generated. Any idea what could be causing this? I have h200 with c400 cam.

**Comment by mediarocketdev** [comment]
I don't have the knowledge to contribute but this would be great as I have 2 c410's and need them to be wireless solar powered units, can't get cables to them. Would be great to run everything through Home Assistant.

**Comment by DamienMrtl** [comment]
> [@sarkinios](https://github.com/sarkinios) child devices through hub will require further changes. I do not own one, someone will need to inspect traffic and find out how the communication is estabilished, then it will be basically the same flow.

Which traffic would you need ? I can try to inspect. Streaming to Google Home app (phone ↔ Hub connection traffic) ?

**Comment by JurajNyiri** [comment]
Traffic on port 443 between phone and camera using the Tapo app, or between phone and tplink servers. You will need to unpin certificate in the apk as well.

**Comment by giuseppe99barchetta** [comment]
Any updates on this? I recently bought a Tapo C410, and the only way I’ve found to get it working is by using go2rtc. However, when I link the camera to Frigate, it drains the battery too quickly.

**Comment by djovic78** [comment]
> Traffic on port 443 and 8800 between phone and camera using the Tapo app, or between phone and tplink servers. You will need to unpin certificate in the apk as well.

@JurajNyiri I tried capturing traffic with wireshark filtering for the camera IP as src and dst whilst streaming in the tapo app. I saw exactly zero packages. Am I missing something? 

Also I have no clue on how to unpin a certificate in the apk.

Can you give me some hints on how to get the right information?

**Comment by djovic78** [comment]
> Any updates on this? I recently bought a Tapo C410, and the only way I’ve found to get it working is by using go2rtc. However, when I link the camera to Frigate, it drains the battery too quickly.

@giuseppe99barchetta 
May I ask how you got it to work with go2rtc? What does your URL look like?
I tried rtsp://<user>:<pass>@<camera_ip>:554/stream1 but I always get a timeout.

**Comment by giuseppe99barchetta** [comment]
> > Any updates on this? I recently bought a Tapo C410, and the only way I’ve found to get it working is by using go2rtc. However, when I link the camera to Frigate, it drains the battery too quickly.
> 
> @giuseppe99barchetta 
> May I ask how you got it to work with go2rtc? What does your URL look like?
> I tried rtsp://<username>:<password>@<camera_ip>:554/stream1 but I always get a timeout.

You can't use rtsp url with this type of camera, try with this configuration:

tapo://CLOUD_PASSWORD@192.168.178.176

Cloud password is you Tapo account password.

**Comment by djovic78** [comment]
> > > Any updates on this? I recently bought a Tapo C410, and the only way I’ve found to get it working is by using go2rtc. However, when I link the camera to Frigate, it drains the battery too quickly.
> > 
> > 
> > [@giuseppe99barchetta](https://github.com/giuseppe99barchetta)
> > May I ask how you got it to work with go2rtc? What does your URL look like?
> > I tried rtsp://:@<camera_ip>:554/stream1 but I always get a timeout.
> 
> You can't use rtsp url with this type of camera, try with this configuration:
> 
> tapo://CLOUD_PASSWORD@192.168.178.176
> 
> Cloud password is you Tapo account password.

Thanks, that worked.
via go2rtc I could also add the stream as a Generic Camera. That's essentially all I wanted. 

Still happy to help with other use cases if I can.

**Comment by LexiB68** [comment]
hello, I have a d230s battery doorbell, and the only "integration" I was able to implement was via samsung smartthings. so at least homeassistant gets notified when the cam KI detects something like a move or person.
unfortunately neither the ring event is passed to HA nor am I able to send a snapshot 🙄
(I have installed the tapo custom camera control integration, but it only shows some sensors for the system, no event, not snapshot, no livestream).
so may I ask to guide me how to integrate via go2rtc?
is this way to be used? https://www.home-assistant.io/integrations/go2rtc/
I really dont understand what this 'guide' means,
I am using HAOS!
so do I have to enter the described section go2rtc in the default config although the header on the page above says there is nothing to do?
and if I have to enter the url ahown from guiseppe, I assume I have to enter the IP of the tapo hub, since the doorbell does not have one (its communicating with the hub, somehow)¡
thank you in advance for your support!
KR, alex.

**Comment by savefolla** [comment]
Any chance we can get the same notifications that we get on the Tapo app on home assistant as well? That would probably unlock a bunch of functionality (for example, trigger an alarm when motion is detected) without needing to stream constantly.

**Comment by algopt** [comment]
Hi! I have a number of TAPO cameras. Most are work wonderfully, but then I have a two battery operated ones, a C425 (that has a solar panel feeding it) and a D230 Doorbell that came with an H200 base. 

Is there anything I can do to help? Like pull some logs? I would love to be able to stream video from these two upon request.

**Comment by JurajNyiri** [comment]
See https://github.com/JurajNyiri/HomeAssistant-Tapo-Control/issues/507#issuecomment-2690995664
https://github.com/JurajNyiri/HomeAssistant-Tapo-Control/issues/507#issuecomment-2706503767

Once I have a bit more time I can continue working on this, what is missing is:
1. Finding a way how to get stream data from a Hub device. I remember reading in some other thread, or possibly here, that the request flow is just a bit different. I have no way to test this though as I do not own such a device.
2. Including what I already have into this integration, just needs some work.

**Comment by Dan1jel** [comment]
How can i add C410? I have controls but no picture at all. My 10y ols Arlo is working but this... ?!?

**Comment by parisdalas** [comment]
I recently install a C425 camera. I didn't search about the non stream problem for home assistant before I buy it so i stuck with that problem. I noticed that the stream via tapo app while is charging with solar is fine and seems to work good but when it gets dark and the  charging stops it is really laggy and almost unable to see the live stream via app. I wonder if I connect it with power supply and after reset the camera, in the installation procces choose install with power supply instead of solar panel if it will makes any difference. Maybe we will able to get stream link with power supply on?

**Comment by apkuki** [comment]
Tapo really did sh... there for his battery operated cameras. Arlo battery have the same problem - here you can make it work https://github.com/kaffetorsk/arlo-streamer by having access via account directly - and if the camera is in battery saving mode the stream just shows a still image and as soon it goes on based on some PIR Sensor etc. it streams the stream based on the camera settings.

That would be nice and everything you actually need... but the seem to give API access only to dedicated partner. I would think HomeAssistant should be in their interest. Because the cameras are actually nice

**Comment by rivergo1d** [comment]
> Hi! I have a number of TAPO cameras. Most are working wonderfully with Home Assistant, but then I have two battery operated ones, a C425 (that has a solar panel feeding it) and a D230 Doorbell that came with an H200 base.
> 
> Is there anything I can do to help? Like pull some logs? I would love to be able to stream video from these two upon request.

Can you tell me, is HA able to detect when your d230 doorbell button is pressed and/or the h200 chime sounds, and then use that as a trigger event for other follow on actions?

**Comment by JurajNyiri** [comment]
**Update:**

I was able to get stream working inside HA, snapshots, mjpeg stream and h264 stream included (no audio yet, having some issues with that...).

In a current state, it is a bit messy, might make HA unstable, and needs a lot of refactoring, but it works! If someone wants to try it out:

1. Enable HD camera entity if you disabled it.
2. Install https://github.com/JurajNyiri/HomeAssistant-Tapo-Control/tree/experiment-pipe-stream manually, it will not work until you do step 3.
3. Grab https://github.com/JurajNyiri/pytapo/tree/experiment-stream-hls, extract pytapo folder (containing [`__init__.py`](https://github.com/JurajNyiri/pytapo/blob/experiment-stream-hls/pytapo/__init__.py)) into custom_components/tapo_control, so that the folder pytapo is next to translations folder.
4. Restart HA

In this version as of today, there is ONLY this new type of stream supported and RTSP was deleted completely. This will of course not be the case in production release.


Optional test: You can disable/enable MJPEG by disabling/enabling "Use Stream from Home Assistant (restart required)" in Device -> Configure -> Configure Device.

Any feedback to this super early version is welcomed.

**I need your help:**

This will not work on cameras that work through a hub. If you have such a camera, I can try getting it to work but will need access to it. If you can port forward and are willing to share access to the device, I can try getting it to work. Reach out to me through Discord.

**Comment by JurajNyiri** [comment]
Update:

More work on the above branches, the stream is now more stable and can recover through restarts or random misc failures.

The delay on mjpeg is pretty amazing, sometimes better compared to webrtc, basically sub 1 second.

**Comment by apkuki** [comment]
Very cool - gonna try this soon. Is there by any chance the possibility of having a self hosted version of just the streamer, to just grab the camera streams without HA and feed them to Frigate or any other NVR

**Comment by JurajNyiri** [comment]
Update:

Code has been significantly refactored and both rtsp and direct stream entities are now available. Direct entities are disabled by default if rtsp ones are available. 

Installation instructions remain the same. 

I still need someone with a hub camera to help me add support for that as well.

**Comment by rivergo1d** [comment]
I have just bought a d230s1 doorbell which uses a h200 hub, if that is if any use?

**Comment by JurajNyiri** [comment]
@rivergo1d yes that would work! If you can port forward and are willing to share your account credentials with me for testing, please reach out to me on Discord (link in readme).

**Comment by rivergo1d** [comment]
Ok, I'm new to this and to discord, so please bear with me. I'll contact you on discord in the next day or so...

**Comment by JurajNyiri** [comment]
Hi @rivergo1d, any luck? Still waiting for a volunteer so that I can finish and release this.

**Comment by rivergo1d** [comment]
Hi, I did message you on discord but heard nothing back "Hi, I offered to lend you access to my h200 and d430s1 doorbell. I've not yet set up Home Assistant for it. When you ask me for my credentials, is that for Tapo or for something else? My hub and camera also have "device sharing" in the Tapo app and I don't know if that also could work"

**Comment by JurajNyiri** [comment]
@rivergo1d My apologies, did not receive notification. I sent you a DM.

**Comment by JurajNyiri** [comment]
Implemented in https://github.com/JurajNyiri/HomeAssistant-Tapo-Control/releases/tag/7.0.0.

**Comment by giuseppe99barchetta** [comment]
> Implemented in https://github.com/JurajNyiri/HomeAssistant-Tapo-Control/releases/tag/7.0.0.

Hi @JurajNyiri, thank you for this release!

I'm having trouble adding a battery-powered camera. I have a Tapo C410 with a solar panel, but when I try to add it to the integration, I get a "Connection failed" error. Could you please let me know which port I should use?

Thanks again!

**Comment by JurajNyiri** [comment]
@giuseppe99barchetta Use port 443 (all default).

1. Make sure camera can be reached from HA
2. Make sure you are entering correct ip
3. Make sure camera is online and responsive, battery devices tend to sleep. Wake it up somehow, through an app for example and keep pressing next button on where you are entering IP.

**Comment by giuseppe99barchetta** [comment]
> [@giuseppe99barchetta](https://github.com/giuseppe99barchetta) Use port 443 (all default).
> 
> 1. Make sure camera can be reached from HA
> 2. Make sure you are entering correct ip
> 3. Make sure camera is online and responsive, battery devices tend to sleep. Wake it up somehow, through an app for example and keep pressing next button on where you are entering IP.

The issue is that the first time I press "Connect", it returns an error.
However, if I press it a second time, it correctly proceeds to the cloud password step.

Thank you!

**Comment by giuseppe99barchetta** [comment]
> [@giuseppe99barchetta](https://github.com/giuseppe99barchetta) Use port 443 (all default).
> 
> 1. Make sure camera can be reached from HA
> 2. Make sure you are entering correct ip
> 3. Make sure camera is online and responsive, battery devices tend to sleep. Wake it up somehow, through an app for example and keep pressing next button on where you are entering IP.

I have one more question: I noticed that there are no motion detection sensors exposed. Is that correct?

**Comment by JurajNyiri** [comment]
@giuseppe99barchetta Correct, there is no hope for those unfortunately. They work through onvif and these cameras do not have RTSP nor onvif. 
From what I understand, the other way these cameras communicate the motion event is that they send them straight to cloud and this integration will always be local only.

**Comment by JurajNyiri** [comment]
For what its worth, there is an issue opened already for that: https://github.com/JurajNyiri/HomeAssistant-Tapo-Control/issues/273

**Comment by giuseppe99barchetta** [comment]
> [@giuseppe99barchetta](https://github.com/giuseppe99barchetta) Correct, there is no hope for those unfortunately. They work through onvif and these cameras do not have RTSP nor onvif. From what I understand, the other way these cameras communicate the motion event is that they send them straight to cloud and this integration will always be local only.

Thank you so much for all your hard work. it's truly appreciated!

**Comment by enys** [comment]
I can confirm this works great, thanks a lot for the work on this !

**Comment by apkuki** [comment]
Amazing! Very cool being able to use the cameras in HA. Is there any possibility using that standalone without HA for Frigate only?

**Comment by JurajNyiri** [comment]
Yes with a slight modification in pytapo library a rtsp restream can be opened. 

Please open PR there.

**Comment by KleinRiese** [comment]
Really amazing, never thought this would be possible. Thank you!

I have 2 C410 and both are working fine.
One of it has weak wifi, so sometimes the streams stops or can't start at all.

I noticed a slighty high cpu usage for about 10 minutes, then a new fffmpeg process starts and the cpu usage is normal.
After a while, I have several ffmpeg processes running (eating up some memory...).

Anything I can do (besides improving my wifi...)?

<img width="1573" height="145" alt="Image" src="https://github.com/user-attachments/assets/cf7da546-13fa-4aa7-816a-b6bc82400248" />

**Comment by JurajNyiri** [comment]
@KleinRiese those ffmpeg processes are (should be) closed after inactivity. If you are watching them, you will have 1 ffmpeg process per stream being watched.

**Comment by KleinRiese** [comment]
Okay, I'll need some more time to check it.
Meanwhile, there is no stream running, synchronization finished a while ago and I have several processes running...

<img width="1580" height="456" alt="Image" src="https://github.com/user-attachments/assets/68331e01-7e4e-4a35-b904-77e24cb3c588" />

**Comment by JurajNyiri** [comment]
That doesn't look right at all. If you can find a way to consistently replicate it, please open a new issue and I will get it fixed.

Please do mind the difference between the two ffmpeg commands, we need to find out which one is causing the issue, they are not the same.
